### ISSUE TITLE
ChangeLog parameter unresolved for ChangeSet id in SQL Formatted ChangeLog

### ISSUE THREAD

**Issue by matteoturra** [liquibase/liquibase#1852]
## Environment
**Liquibase Version**: 4.3.x 

## Description
ChangeLog parameter are not expanded (resolved) in ChangeSet id in SQL Formatted ChangeLog.

## Steps To Reproduce
Define a ChangeLog _changelog.xml_with a property and include the same changeset written in XML format and another one written in SQL Format

```xml
<?xml version="1.0" encoding="UTF-8"?>  
<databaseChangeLog  
  xmlns="http://www.liquibase.org/xml/ns/dbchangelog"  
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"  
  xmlns:pro="http://www.liquibase.org/xml/ns/pro"  
  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.2.xsd">
    <property  name="myprop"  value="myval"/>
    <include file="test.sql" relativeToChangelogFile="true"/>
    <include file="test.xml" relativeToChangelogFile="true"/>
</databaseChangeLog>
```

test.sql:

```sql
--liquibase formatted sql
--changeset mturra:ins_${myprop}

select 1 from dual;
```

test.xml

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
    <changeSet id="ins_${myprop}" author="mturra">
        <sql>
            select 1 from dual
        </sql>
    </changeSet>
</databaseChangeLog>
```

Launch liquibase update.

## Actual Behavior
On DATABASECHANGELOG I get two row:

* the row with filename text.xml have id value "ins_myval"
* the row with filename text.sql have id value "ins_${myprop}"

## Expected/Desired Behavior
The same behavior for changeset written in different format.

**Comment by matteoturra** [comment]
Patch:
```
Index: liquibase-core/src/main/java/liquibase/parser/core/formattedsql/FormattedSqlChangeLogParser.java
<+>UTF-8
===================================================================
diff --git a/liquibase-core/src/main/java/liquibase/parser/core/formattedsql/FormattedSqlChangeLogParser.java b/liquibase-core/src/main/java/liquibase/parser/core/formattedsql/FormattedSqlChangeLogParser.java
--- a/liquibase-core/src/main/java/liquibase/parser/core/formattedsql/FormattedSqlChangeLogParser.java	(revision 10a0c509c59401e9351cf7d134b4bc671b5d500f)
+++ b/liquibase-core/src/main/java/liquibase/parser/core/formattedsql/FormattedSqlChangeLogParser.java	(date 1621504467566)
@@ -207,10 +207,11 @@
                        logicalFilePath = changeLog.getLogicalFilePath ();
                     }
                     String dbms = parseString(dbmsPatternMatcher);
-
+                    String changeSetId = changeLogParameters.expandExpressions(changeSetPatternMatcher.group(2), changeLog);
+                    String author = changeSetPatternMatcher.group(1);
 
                     changeSet =
-                       new ChangeSet(changeSetPatternMatcher.group(2), changeSetPatternMatcher.group(1), runAlways, runOnChange, logicalFilePath, context, dbms, runWith, runInTransaction, changeLog.getObjectQuotingStrategy(), changeLog);
+                       new ChangeSet(changeSetId, author, runAlways, runOnChange, logicalFilePath, context, dbms, runWith, runInTransaction, changeLog.getObjectQuotingStrategy(), changeLog);
                     changeSet.setLabels(new Labels(labels));
                     changeSet.setFailOnError(failOnError);
                     changeLog.addChangeSet(changeSet);
@@ -363,4 +364,4 @@
         }
         return resourceAsStream;
     }
-}
+}
```
\ No newline at end of file

**Comment by molivasdat** [comment]
Hi @matteoturra Thanks for creating this issue and a PR to fix it.  We will add it to the list of issues that we are processing.

**Comment by molivasdat** [comment]
Hi @matteoturra Just curious of the use case for what you are trying to accomplish for dynamically setting the changeset id.
I understand that it works for xml but not for a formatted sql changeset which is why this issue is highlighted.  I'd like to take a step back and understand what your requirements are.

**Comment by matteoturra** [comment]
Hi, this is my use case. I need to launch the same sql script on the same
oracle user (schema) from different instances. Usually I run liquibase with
a connection to user A to give grants to user oracle B. The same script run
to give the same grants to user oracle C and so on....
The script has the oracle user to give the grant to in a parameter
${targetUsername} and I use this parameter in the changeset ID too, so
liquibase can tell script running one from the other instance.

Do you have any other suggestions to manage this scenario?

Thanks
Matteo



Il giorno mer 26 mag 2021 alle ore 19:19 Mike Olivas <
***@***.***> ha scritto:

**Comment by geocyp86** [comment]
@molivasdat  @matteoturra  i have the same need.
Basically I need to deploy the same code in the same database under a test client. So Client A and Client B will have 2 separate sets of deployed changes coming from the same code base.
This is to allow some beta functionality to go in the second implementation without keeping + synchronizing the main code.
I ve tested the XML changeset and it is doing what i need but all my files (approx 3000 files) are in SQL format.

I ve tested in 3.4.2 , 4.3.4 and 4.5.0 and it works in all three for XML but does not work in any for SQL.

When should we expect a fix in the SQL format ?

**Comment by mar-deb** [comment]
Hi, do you plan to fix it? This bug blocks the full implementation of liquibase in my organization. 

My use case is that i have one file, and  many scenario to deploy the file. I need to set param depends on scenario.

**Comment by kataggart** [comment]
@mar-deb Apologies for delays on this; we will review with our team. Stay tuned.

**Comment by nvoxland** [comment]
This should have been resolved in v4.7.1. What version are you seeing problems with, @mar-deb ?

**Comment by kataggart** [comment]
closed for now; please re-open if needed thanks
